python/src/trezorlib/authentication.py

```diff
@@ -0,0 +1,306 @@
+        r, s = Certificate._decode_signature_permissive(sig_bytes)
+        reencoded = utils.encode_dss_signature(r, s)
+        if reencoded != sig_bytes:
+            LOG.info(
+                "Re-encoding malformed signature: %s -> %s",
+                sig_bytes.hex(),
+                reencoded.hex(),
+            )
+        return reencoded
+
+
+def verify_authentication_response(
+    challenge: bytes,
+    signature: bytes,
+    cert_chain: t.Iterable[bytes],
+    *,
+    whitelist: t.Collection[bytes] | None,
+    allow_development_devices: bool = False,
+    root_pubkey: bytes | ec.EllipticCurvePublicKey | None = None,
+) -> None:
+    """Evaluate the response to an AuthenticateDevice call.
+
+    Performs all steps and logs their results via the logging facility. (The log can be
+    accessed via the `LOG` object in this module.)
+
+    When done, raises DeviceNotAuthentic if the device is not authentic.
+
+    The optional argument `root_pubkey` allows you to specify a root public key either
+    as an `ec.EllipticCurvePublicKey` object or as a byte-string representing P-256
+    public key.
+    """
+    if isinstance(root_pubkey, (bytes, bytearray, memoryview)):
+        root_pubkey = ec.EllipticCurvePublicKey.from_encoded_point(
+            ec.SECP256R1(), root_pubkey
+        )
+
+    challenge_bytes = (
+        len(CHALLENGE_HEADER).to_bytes(1, "big")
+        + CHALLENGE_HEADER
+        + len(challenge).to_bytes(1, "big")
+        + challenge
+    )
+
+    cert_chain_iter = iter(cert_chain)
+
+    failed = False
+
+    try:
+        cert = Certificate(next(cert_chain_iter))
+    except Exception:
+        LOG.error("Failed to parse device certificate.")
+        raise DeviceNotAuthentic
+
+    try:
+        cert.verify(signature, challenge_bytes)
+    except exceptions.InvalidSignature:
+        LOG.error("Challenge verification failed.")
+        failed = True
+    else:
+        LOG.debug("Challenge verified successfully.")
+
+    cert_label = "Device certificate"
+    for i, issuer_bytes in enumerate(cert_chain_iter, 1):
+        try:
+            ca_cert = Certificate(issuer_bytes)
+        except Exception:
+            LOG.error(f"Failed to parse CA certificate #{i}.")
+            failed = True
+            continue
+
+        if whitelist is None:
+            LOG.warning("Skipping public key whitelist check.")
+        else:
+            if ca_cert.public_key_bytes() not in whitelist:
+                LOG.error(f"CA certificate #{i} not in whitelist: %s", ca_cert)
+                failed = True
+
+        try:
+            cert.verify_issued_by(ca_cert)
+        except exceptions.InvalidSignature:
+            LOG.error(f"{cert_label} verification failed: %s", cert)
+            failed = True
+        else:
+            LOG.debug(f"{cert_label} verified successfully: %s", cert)
+        finally:
+            cert = ca_cert
+            cert_label = f"CA #{i - 1} certificate"
```

Comment:
This seems to me like it should be `i`, not `i-1`. On line 223 we refer to the same certificate by index `i`, don't we?

Reply:
indeed, this seems to be a leftover from some other version of this code
5623f6d6f7f5a8a1e49b73959117d1ec499bc579